### PR TITLE
⚡ Re-wrote chunk parser to go faster

### PR DESCRIPTION
**Pull request by chrissorchard**
So much faster. But also beware slightly questionable python.

### CODE REVIEW

File: pink_concrete/__init__.py [hunk 1]
```diff
@@ -18,88 +19,68 @@
...
-AIR_CHUNK = _air_chunk()
-
-
-def _chunk_dict(chunk: 'Chunk'):
-    cdict = {}
-    y = 0
-    z = 0
-    x = 0
-    for section_index in range(16):
+    for section_index in reversed(range(16)):
```
Comment (coiax) [inline pink_concrete/__init__.py]: So fun fact, I used the `timeit` module to compare the speeds of `reversed(range(16))` vs `range(15, -1, -1)`. About a 100ns difference, which would add up to about 0.31 seconds across the entire agnomen map; so not really worth losing the readability. :sweat_smile:

Comment (chrissorchard) [reply pink_concrete/__init__.py]: but those precious ns 😛

File: pink_concrete/__init__.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
 import typing
 import multiprocessing
 import random
+from collections import defaultdict
```
Comment (coiax) [inline pink_concrete/__init__.py]: Probably can just import the collections module, and go `collections.defaultdict` elsewhere.

File: pink_concrete/__init__.py [hunk 3]
```diff
@@ -18,88 +19,68 @@
...
-def top_down_until(chunkdict, x: int, z: int) -> typing.Iterator['Block']:
-    """Return an iterator of blocks in the chunk at x,z
-    until the returned block is opaque, or there are no blocks left."""
-
-    assert 0 <= x <= 15
-    assert 0 <= z <= 15
+def get_chunk_stacks(
+    chunk: 'Chunk'
+)-> typing.Dict[typing.Tuple[int,int], typing.Iterator['Block']]:
+    """For a given chunk genrate a dict of translucent and the
```
Comment (coiax) [inline pink_concrete/__init__.py]:
```suggestion
    """For a given chunk generate a dict of translucent and the
```

File: pink_concrete/__init__.py [hunk 4]
```diff
@@ -18,88 +19,68 @@
...
+    top opaque blocks for each of the internal x,z pairs"""
+    #default dict that lets us skip completed block stacks
```
Comment (coiax) [inline pink_concrete/__init__.py]: Put a space between your comment and the #

File: pink_concrete/__init__.py [hunk 5]
```diff
@@ -18,88 +19,68 @@
...
-
-    chunk_dict = _chunk_dict(chunk)
-
-    for x in range(16):
-        for z in range(16):
-            yield x, z, top_down_until(chunk_dict, x, z)
+                    local_y += 1
 
+        #add section to total stack
+        for x,z in section_stacks:
```
Comment (coiax) [inline pink_concrete/__init__.py]:
```suggestion
        for x, z in section_stacks:
```

File: pink_concrete/__init__.py [hunk 6]
```diff
@@ -18,88 +19,68 @@
...
+            chunk_stacks[x,z].extend(reversed(section_stacks[x,z]))
```
Comment (coiax) [inline pink_concrete/__init__.py]:
```suggestion
            chunk_stacks[x, z].extend(reversed(section_stacks[x, z]))
```

**Pull request merged**